readme.md
# Queue Shedule

[![NPM version](https://img.shields.io/npm/v/queue-schedule.svg?style=flat-square)](https://www.npmjs.com/package/queue-schedule)
[![Build status](https://travis-ci.com/yunnysunny/queue-schedule.svg?branch=master)](https://travis-ci.com/yunnysunny/queue-schedule)

Kafka is a high avaliable message queue, but it lacks of consuming message with a slow speed. Some of task with no need to finish it at none, and we want to complete it with a small cost. This is just the reason why we develop `Queue Shedule`.

## Install

```npm install queue-schedule```

## How to use

### Use kafka-node
A basic example is showed as follows:

```javascript
const kafka = require('kafka-node');
const {expect} = require('chai');
const {SHKafkaProducer,SHKafkaConsumer} = require('queue-schedule');
const KAFKA_HOST = process.env.KAFKA_PEERS;
const FIST_DATA = {a:1,b:2};
const SCHEDULE_NAME1 = 'schedule1';
const TOPIC_NAME1 = 'topic.1';
const client = new kafka.KafkaClient({kafkaHost: KAFKA_HOST});
const producerCreater = new kafka.HighLevelProducer(client);
const producerPromise = new Promise(function(resolve,reject) {
    producerCreater.on('ready',function() {
        resolve(producerCreater);
    });

    producerCreater.on('error',function(err) {
        reject(err);
    });
});

let hasDone = false;
new SHKafkaConsumer({
    name: 'kafka',
    topics:[TOPIC_NAME1],
    consumerOption: {
        kafkaHost: KAFKA_HOST,
        fromOffset: 'earliest',
        fetchMaxBytes: 1024*1024,
    },
    doTask:function(messages,callback) {console.log(messages);
        if (!hasDone) {
            const value = messages[0].value;
            let data = null;
            try {
                data = JSON.parse(value);
            } catch (e) {
                hasDone = true;
                console.error('parse message error',e);
                return;
            }
            expect(data).to.have.property('a').and.equal(1);
            console.log('recieve data',data);
            hasDone = true;
        }
        callback();
    },
    readCount : 1,
    pauseTime : 500,
    idleCheckInter: 10 * 1000
}).on(KafkaConsumer.EVENT_CONSUMER_ERROR,function(err) {
    console.error('consumer error',err);
    hasDone = true;
    
});


new SHKafkaProducer({
    name : SCHEDULE_NAME1,
    topic: TOPIC_NAME1,
    producerPromise
}).addData(FIST_DATA,{},function(err) {
    if (err) {
        console.error('write to queue error',err);
        return ;
    }
    console.info('write to kafka finished');
});
```

### Use rdkafka

```javascript
const Kafka = require('node-rdkafka');
const {RdKafkaProducer,RdKafkaConsumer} = require('queue-schedule');
const producerRd = new Kafka.HighLevelProducer({
    'metadata.broker.list': KAFKA_HOST,
    'linger.ms':0.1,
    'queue.buffering.max.ms': 500,
    'queue.buffering.max.messages':1000,
    // debug: 'all'
});
producerRd.on('event.error',function(err) {
    slogger.error('producer error');
});
producerRd.on('event.log',function(log) {
    slogger.debug('producer log',log);
});
const producer = new RdKafkaProducer({
    name : SCHEDULE_NAME1,
    topic: TOPIC_NAME1,
    producer:producerRd,
    delayInterval: 500
});
producer.addData(FIST_DATA, {},function(err) {
    if (err) {
        slogger.error('write to queue error',err);
        return done('write to queue error');
    }
    slogger.info('write to kafka finished');
});


const consumer = new Kafka.KafkaConsumer({
    'metadata.broker.list': KAFKA_HOST,
    'group.id': 'test-rdkafka-0',
    'auto.offset.reset':'earliest',
    'socket.keepalive.enable': true,
    'socket.nagle.disable': true,
    'enable.auto.commit': true,
    'fetch.wait.max.ms': 5,
    'fetch.error.backoff.ms': 5,
    'queued.max.messages.kbytes': 1024 * 10,
    debug:'all'
});
let hasDone = false;
new RdKafkaConsumer({
    name: 'kafka',
    consumer,
    topics: [ TOPIC_NAME1],
    
    doTask:function(messages,callback) {
        slogger.trace(messages);
    },
    readCount : 1,
    pauseTime : 500,
    idleCheckInter: 10 * 1000
}).on(RdKafkaConsumer.EVENT_CONSUMER_ERROR,function(err) {
    slogger.error('consumer error',err);
    hasDone = true;
    done(err);
}).on(RdKafkaConsumer.EVENT_CLIENT_READY,function() {
    slogger.trace('the consumer client is ready');
    
}).on(RdKafkaConsumer.EVENT_LOG,function(log) {
    // slogger.trace(JSON.stringify(log));
});
```

### Using kafkajs

```javascript
const { Kafka } = require('kafkajs');
const {KafkaJsProducer,KafkaJsConsumer} = require('queue-schedule');

const FIST_DATA = {a:1,b:2};
const SCHEDULE_NAME1 = 'schedule1';
const TOPIC_NAME1 = 'topic.kafkajs';
const client =  new Kafka({
    brokers: ['xxxx', 'yyyy']
});

const producer = new KafkaJsProducer({
    topic: TOPIC_NAME1,
    client,
});
producer.addData(FIST_DATA, {},function(err) {
    if (err) {
        console.error('write to queue error',err);
        return;
    }
    console.info('write to kafka finished');
});
producer.on(KafkaJsProducer.EVENT_PRODUCER_ERROR, function(err) {
    console.error('error in consumer', err);
});

new KafkaJsConsumer({
    name: 'kafka',
    client,
    topic: TOPIC_NAME1,
    consumerOption: {
        groupId: 'kafkajs',
        fromBeginning: true
    },
    doTask:function(messages,callback) {
        console.log(messages);
        const value = messages[0].value;//read the first value
        let data = null;
        try {
            data = JSON.parse(value);
            console.log('recieve data',data);
        } catch (e) {
            console.error('parse message error',e);
        }

        callback();//the next loop
    },
    readCount : 1,
    pauseTime : 500,
    idleCheckInter: 10 * 1000
}).on(KafkaJsConsumer.EVENT_CONSUMER_ERROR,function(err) {
    console.error('consumer error',err);
    hasDone = true;
    done(err);
}).on(KafkaJsConsumer.EVENT_CONSUMER_READY,function() {
    console.log('the consumer is ready');
});
```
## API

For detail usage, see the document online [here](https://yunnysunny.github.io/queue-schedule).

## Known issue

1. For the library of `kafka-node` is not stable, we suggest you using the `rdkafka` library. In othter words, you'd better use `RdKafkaProducer` and `RdKafkaConsumer` instead of `SHKafkaProducer` and `SHKafkaConsumer`.

## License

[MIT](LICENSE)


CHANGELOG.md
# v2.5.1

## Fix

1. Fix the issue of not calling callback function when addData return false in the class of `RdKafkaProducer` `SHKafkaProducer` `KafkaJsProducer`.

# v2.5.0

## Add

1. Add the feature of encoding data manual, see `RdKafkaProducer` and `KafkaJsProducer`.

# v2.4.0
## Add
1. Add the support of kafkajs, see `KafkaJsProducer` and `KafkaJsConsumer`.

# v2.3.5
## Fix
1. Fix the memory leak when the connection of broker is failed in `RdKafkaProducer`.
2. Fix the the broken when sending message while the connection is not ready in `SHKafkaProducer`.

# v2.3.4
## Fix
1. Remove the cache data in memory when use rdkafka.

# v2.3.3
## Fix
1. Just fix the building error of travis-ci.

# v2.3.2
## Fix
1. Fix the problem of not sending data in given delay time.

# v2.3.1
## Improvement
1. Remove the flush operation in `RdKafkaProducer`.

# v2.3.0
## Improvement
1. Set the timestamp to `0` when call rdkafka's produce function.

# v2.2.1
## Fix
1. It may case to memory leak when using `RdKafkaConsumer`, but I can't reproduce it, just add some code to defend it.

# v2.2.0
## Breaking Changes
1. Rename the `KafkaProducer` to `SHKafkaProducer`, rename the `KafkaConsumer` to `SHKafkaConsumer`, drop the object of `globalEvent`.  
2. If you still use the libary of `kafka-node`, you have to wrap an object of `Promise` and give the value of an instance of `HighLevelProducer` when the instance is ready. For example:

```javascript
const client = new kafka.KafkaClient({kafkaHost: 'the peers of you kafka cluster'});
const producerCreater = new kafka.HighLevelProducer(client);
const producerPromise = new Promise(function(resolve,reject) {
    producerCreater.on('ready',function() {
        resolve(producerCreater);
    });

    producerCreater.on('error',function(err) {
        reject(err);
    });
});
const producerSH = new SHKafkaProducer({
    name : 'myname',
    topic: 'mytopic',
    producerPromise
});
```

# v2.1.0
## Add
1. Add the choice of the driver of `node-librakafka` to produce and consume message, see `RdKafkaProducer` and `RdKafkaConsumer`.

# v2.0.0
## Improvement
1. Bump `kafka-node` to 5.0.0.

# v1.1.1
## Improvement
1. Bump `kafka-node` to 4.1.3.

# v1.1.0
## Add
1. Add the object of `globalEvent` to emit the events of client's and producer's status.
2. Use a producer pool to cache the producer with same kafka host, in other words, you will only create a single producer instance, when you instantiate diffrent `KafkaProducer` with the same kafka host.

## Remove
1. Remove the event of `EVENT_CLIENT_READY` `EVENT_CLIENT_ERROR` `EVENT_CLIENT_CLOSE` `EVENT_PRODUCER_READY` `EVENT_PRODUCER_ERROR` from `KafkaProducer`, these events mentioned above, are move to the object `globalEvent`.

# v1.0.0
## Add
1. Emit event when kafkaClient closed.
2. Add the parameter of `prepareMiddleware` to `KafkaProducer`.

## Breaking Changes
1. Remove the option of `zookeeperHost` form `KafkaProducer` and `KafkaConsumer`.
2. Remove the `manager` object.
3. Remove the function of `addDataDelay` form `KafkaProducer`.
4. Remove the parameter of `topicList` from `KafkaProducer`.
5. The function `addData` of `KafkaProducer` returns a boolean value, which indicate whether the data to send is vaild.

# v0.6.0
## Add
1. Bump the package `kafka-node` to 3.0.1.

# v0.5.3
## Fix
1. Resovle the issue of double stringify.

# v0.5.2
## Add
1. Add parameter of `fromOffset` to KafkaConsumer.

# v0.5.1
## Add
1. Add feature of emitting event of `KafkaProducer.EVENT_SEND_ERROR`.

# v0.5.0
## Add
1. Use consumerGroup instead of consumer.
2. Add the feature of using broker url to connect to the kafka client.

# v0.3.1
## Add
1. Use HighLevelProducer instead of Producer.
2. Add the feature of setting the groupId of consumer.
3. Add the parameter of the count of messages published to the event of `EVENT_DELAY_MESSAGE_SEND_FINISHED`

# v0.2.1
## Add
1. Do code refactoring, create kafka client in the KafkaProducer and KafkaConsumer.
2. Add feature of emitting event of `KafkaProducer.EVENT_CLIENT_READY` `KafkaProducer.EVENT_CLIENT_ERROR` `KafkaProducer.EVENT_PRODUCER_READY` `KafkaProducer.EVENT_PRODUCER_ERROR` `KafkaConsumer.EVENT_CLIENT_READY` `KafkaConsumer.EVENT_CLIENT_ERROR` `KafkaConsumer.EVENT_CONSUMER_ERROR`.
## Fix
1. Fxed the error of breaking down when call addKafkaSchedule twice with the same name.

# v0.2.0
## Add
1. Add some testcases.
2. Add the feature of emitting event of  `KafkaProducer.EVENT_DELAY_MESSAGE_SEND_FINISHED` when delay message publish finished.

# v0.1.0
## Add
1. Add feature f sending message delay.
2. Add travis-ci.yml.



lib/KafkaConsumer.js
const EventEmitter = require('events');
const kafka = require('kafka-node');
/**
 * @callback DoTask
 * 
 * @param {Object[]} messages
 */


/**
 * @typedef ConsumerOption
 * 
 * The options passed to [kafka.ConsumerGroup](https://github.com/SOHU-Co/kafka-node#consumergroup) 
 * 
 * @param {String=} kafkaHost connect directly to kafka broker (instantiates a KafkaClient), when not passed, it will use KafkaConsumerOption.kafkaHost in default.
 * @param {Boolean} [ssl=false] optional (defaults to false) or tls options hash.
 * @param {String=} [encoding='utf8'] default is utf8, use 'buffer' for binary data.
 * @param {String=} groupId it will use KafkaConsumerOption.groupId in default.
 * @param {Number=} [fetchMaxBytes=1024*1024]
 * @param {Number} [sessionTimeout=15000]
 * @param {String[]} [protocol=['roundrobin']] An array of partition assignment protocols ordered by preference. 'roundrobin' or 'range' string for built ins
 * @param {String} [fromOffset=earliest] Offsets to use for new groups other options could be 'earliest' or 'none' (none will emit an error if no offsets were saved) quivalent to Java client's auto.offset.reset
 * @param {String} [outOfRangeOffset=earliest] how to recover from OutOfRangeOffset error (where save offset is past server retention) accepts same value as fromOffset
 * @param {Boolean} [migrateHLC=false]
 * @param {Boolean} [migrateRolling=true]
 */

/**
 * @typedef KafkaConsumerOption
 * 
 * @param {String} name The name of current instance.
 * @param {String=} kafkaHost The host of the broker of kafka.
 * @param {String[]} topics The topics that will be consumed.
 * @param {ConsumerOption=} consumerOption The option to create a new instance of `Kafka.ConsumerGroup`.
 * @param {Number} readCount After reading the count of `readCount`, the consumer will be paused.
 * @param {Number} pauseTime The duration of pause time, after that the consumer will be continued.
 * @param {DoTask} doTask The consume process function.
 * @param {Number} idleCheckInter The instance of KafkaConsumer has a timer inner, to check whether the process of `doTask` is idle. The timer will trigger every `idleCheckInter` ms. 
 */

/**
 * The class of KafkaConsumer
 * @class KafkaConsumer
 * @extends {EventEmitter}
 */

class KafkaConsumer extends EventEmitter {
    
    /**
     * Creates an instance of KafkaConsumer. It will call the function of #consumer inner.
     * 
     * @param {KafkaConsumerOption} option
     * @memberof KafkaConsumer
     */
    constructor({
        name,
        kafkaHost,
        topics,
        consumerOption = {},
        doTask,
        readCount = 100,
        pauseTime = 500,
        idleCheckInter = 1000 * 10,
        
    }) {
        super();
        this.name = name;
        this.readCount = readCount;
        this.consumer = null;
        if (!topics ||  !consumerOption) {
            throw new Error('The parameters of  topics and consumerOption must be given.');
        }
        if (!kafkaHost && !consumerOption.kafkaHost) {
            throw new Error('The kafka host must be given in kafkaHost or in consumerOption.kafkaHost.');
        }
        consumerOption.groupId = consumerOption.groupId || name;
        consumerOption.kafkaHost = consumerOption.kafkaHost || kafkaHost;
        consumerOption.fromOffset =  consumerOption.fromOffset || 'earliest';
        this._init(topics,consumerOption);
        this.pauseTime = pauseTime;
        this.doTask = doTask;
        this.lastFinishedTime = 0;
        this.idleCheckInter = idleCheckInter;
        this.messages = [];
        this.consume(doTask);
    }

    _init(topics,consumerOption) {
        // const client = new kafka.Client(zookeeperHost);
        const _this = this;
        // client.on('ready',function() {
        //     _this.emit(KafkaConsumer.EVENT_CLIENT_READY);
        // });
        // client.on('error',function(err) {
        //     _this.emit(KafkaConsumer.EVENT_CLIENT_ERROR,err);
        // });
        const names = [];
        for (const topic of topics) {
            if (typeof (topic) === 'string') {
                names.push(topic);
            } else {
                names.push(topic.topic);
            }
        }
        const consumer = this.consumer = new kafka.ConsumerGroup(
            consumerOption,names
        );
        consumer.on('error',function(err) {
            _this.emit(KafkaConsumer.EVENT_CONSUMER_ERROR,err);
        });
    }

    _continue() {
        const _this = this;
        setTimeout(function resume() {
            _this.consumer.resume();
        }, this.pauseTime);
    }
    /**
     * The consume function.
     * Do not call this function manual!
     * 
     * @param {DoTask} doTask 
     * @memberof KafkaConsumer
     */
    consume(doTask) {
        let messages = this.messages;
        const consumer = this.consumer;
        const count = this.readCount;
        const _this = this;

        consumer.on('message', function(message) {
            if (message.offset % count === 0) {
                consumer.pause();
                messages.push(message);
                _this.lastFinishedTime = new Date().getTime();
                doTask(messages.splice(0), function () {
                    _this._continue();
                });
            } else {
                messages.push(message);
                _this.lastFinishedTime = new Date().getTime();
            }
        });

        setInterval(function clear() {
            const idle = new Date().getTime() - _this.lastFinishedTime;
            if ( idle > 1000 && messages.length > 0) {
                consumer.pause();
                doTask(messages.splice(0), function() {
                    _this._continue();
                });
            }
        }, this.idleCheckInter);
    }
}
/**
 * The event to notify that the client is ready.
 */
KafkaConsumer.EVENT_CLIENT_READY = 'eventClientReady';
/**
 * The event to notify that the client is error.
 */
KafkaConsumer.EVENT_CLIENT_ERROR = 'eventClientError';
/**
 * The event to notify that an error ocurred in consumer.
 */
KafkaConsumer.EVENT_CONSUMER_ERROR = 'eventConsumerError';

module.exports = KafkaConsumer;
